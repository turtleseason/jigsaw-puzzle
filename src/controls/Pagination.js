import { Component } from 'react';
import PropTypes from 'prop-types';


export default class Pagination extends Component {
    handleNavigate(e, newPage) {
        e.preventDefault();
        if (this.props.onPageChange) {
            this.props.onPageChange(newPage);
        }
    }

    renderPageLinks() {
        const links = [];

        const start = this.props.currentPage - this.props.numPageLinks - 1;
        const end = this.props.currentPage + this.props.numPageLinks + 1;

        if (start >= this.props.minPage) {
            links.push(<li key={start}><span className='btn' style={{ cursor: 'default' }}>...</span></li>);
        } else {
            links.push(<li key={start} className='invisible'></li>);
        }

        for (let i = start + 1; i < end; i++) {
            let content = null;
            if (i === this.props.currentPage) {
                content = <span className='btn' style={{ cursor: 'default' }}>{i}</span>;
            }
            else if (i >= this.props.minPage && i <= this.props.maxPage) {
                content = <button className='btn btn-link' type='button' onClick={(e) => this.handleNavigate(e, i)}>{i}</button>;
            }
            links.push(<li className={!content ? 'invisible' : null} key={i}>{content}</li>);
        }

        if (end <= this.props.maxPage) {
            links.push(<li key={end}><span className='btn' style={{ cursor: 'default' }}>...</span></li>);
        } else {
            links.push(<li key={end} className='invisible'></li>);
        }

        return <ul>{links}</ul>;
    }

    render() {
        const { currentPage, minPage, maxPage, numPageLinks, onPageChange, className, ariaLabel, ...customAttributes } = this.props;

        const outerClassName = 'custom-pagination container ' + (className || '');
        const outerAriaLabel = ariaLabel || 'Search results pages';

        const isFirst = !(this.props.currentPage > this.props.minPage);
        const isLast = !(this.props.currentPage < this.props.maxPage);

        const pageLinks = this.renderPageLinks();

        return (
            <nav className={outerClassName} aria-label={outerAriaLabel} {...customAttributes}>
                <div className='row justify-content-between align-items-center'>
                    <div className='col-auto p-0'>
                        <button className={'btn btn-link' + (isFirst ? ' invisible' : '')} type='button'
                            onClick={isFirst ? null : (e) => this.handleNavigate(e, this.props.currentPage - 1)}>
                            {'< Back'}
                        </button>
                    </div>

                    <div className='col-auto'>
                        {pageLinks}
                    </div>

                    <div className='col-auto p-0'>
                        <button className={'btn btn-link' + (isLast ? ' invisible' : '')} type='button'
                            onClick={isLast ? null : (e) => this.handleNavigate(e, this.props.currentPage + 1)}>
                            {'More >'}
                        </button>
                    </div>
                </div>
            </nav>
        );
    }
}

// In addition to the listed types, any unrecognized props will be added as attributes to the outer <nav> DOM node,
// to allow for customization
Pagination.propTypes = {
    currentPage: PropTypes.number.isRequired,
    minPage: PropTypes.number,
    maxPage: PropTypes.number,
    numPageLinks: PropTypes.number,  // The number of page links shown on either side of the current page number
    onPageChange: PropTypes.func,    // Callback raised when a page link is clicked; it is given one parameter, the page number to navigate to
    className: PropTypes.string,     // An optional class string to append to the default classes
    ariaLabel: PropTypes.string,     // An optional label string to replace the default aria-label
};

Pagination.defaultProps = {
    minPage: 1,
    maxPage: Number.MAX_SAFE_INTEGER,
    numPageLinks: 1,
};