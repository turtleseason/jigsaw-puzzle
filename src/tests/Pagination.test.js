import { render, screen, getByRole, getByText } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import '@testing-library/jest-dom';

import Pagination from '../controls/Pagination.js';


describe('Pagination', () => {
    test('renders without exploding', () => {
        render(<Pagination currentPage={3}/>);
    });

    test('renders correctly in middle of page range', () => {
        const { asFragment } = render(<Pagination currentPage={5} numPageLinks={2}/>);
        expect(asFragment()).toMatchSnapshot();
    });

    // (For comparison of testing methods - manually checking for important elements vs. snapshot testing)
    test('renders correctly in middle of page range (non-snapshot)', () => {
        render(<Pagination currentPage={5} numPageLinks={2}/>);

        expect(screen.getByRole('link', {name : '< Back'})).toBeVisible();
        expect(screen.getByRole('link', {name : 'More >'})).toBeVisible();
        
        const expectedItems = ['...', '3', '4', '5', '6', '7', '...'];
        
        const listItems = screen.getAllByRole('listitem');
        expect(listItems).toHaveLength(expectedItems.length);
        
        for (let i = 0; i < listItems.length; i++) {
            const content = expectedItems[i];
            if (content === '...' || content === '5') {
                expect(getByText(listItems[i], content)).toBeInTheDocument();
            } else {
                expect(getByRole(listItems[i], 'link', { name: content })).toBeInTheDocument();
            }
        }
    });
    
    test('renders correctly at start of page range', () => {
        const { asFragment } = render(<Pagination currentPage={1} minPage={1} numPageLinks={2}/>);
        expect(asFragment()).toMatchSnapshot();
    });
    
    test('renders correctly at end of page range', () => {
        const { asFragment } = render(<Pagination currentPage={10} maxPage={10}/>);
        expect(asFragment()).toMatchSnapshot();
    });

    test('renders correctly for whole page range', () => {
        const { asFragment } = render(<Pagination currentPage={2} minPage={1} maxPage={4} numPageLinks={2}/>);
        expect(asFragment()).toMatchSnapshot();
    });

    test('"back" callback works', () => {
        const mockFn = jest.fn();
        
        render(<Pagination currentPage={5} minPage={1} maxPage={10} onPageChange={mockFn}/>);

        userEvent.click(screen.getByRole('link', { name: /Back/ }));
        expect(mockFn).toHaveBeenCalledWith(4);
    });

    test('"next" callback works', () => {
        const mockFn = jest.fn();

        render(<Pagination currentPage={5} minPage={1} maxPage={10} onPageChange={mockFn}/>);

        userEvent.click(screen.getByRole('link', { name: /More/ }));
        expect(mockFn).toHaveBeenCalledWith(6);
    });

    test('number link callbacks work', () => {
        const mockFn = jest.fn();

        render(<Pagination currentPage={5} numPageLinks={3} onPageChange={mockFn}/>);

        for (let i = 2; i <= 8; i++) {
            if (i !== 5) {  // skip current page (it is not a link)
                userEvent.click(screen.getByRole('link', { name: i }));
                expect(mockFn).toHaveBeenLastCalledWith(i);        
            }
        }
    });
});