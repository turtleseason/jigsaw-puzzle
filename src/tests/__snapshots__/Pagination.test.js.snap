// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination renders correctly at end of page range 1`] = `
<DocumentFragment>
  <nav
    aria-label="Search results pages"
    class="custom-pagination container "
  >
    <div
      class="row justify-content-between align-items-center"
    >
      <div
        class="col-auto p-0"
      >
        <button
          class="btn btn-link"
          type="button"
        >
          &lt; Back
        </button>
      </div>
      <div
        class="col-auto"
      >
        <ul>
          <li>
            <span
              class="btn"
              style="cursor: default;"
            >
              ...
            </span>
          </li>
          <li>
            <button
              class="btn btn-link"
              type="button"
            >
              9
            </button>
          </li>
          <li>
            <span
              class="btn"
              style="cursor: default;"
            >
              10
            </span>
          </li>
          <li
            class="invisible"
          />
          <li
            class="invisible"
          />
        </ul>
      </div>
      <div
        class="col-auto p-0"
      >
        <button
          class="btn btn-link invisible"
          type="button"
        >
          More &gt;
        </button>
      </div>
    </div>
  </nav>
</DocumentFragment>
`;

exports[`Pagination renders correctly at start of page range 1`] = `
<DocumentFragment>
  <nav
    aria-label="Search results pages"
    class="custom-pagination container "
  >
    <div
      class="row justify-content-between align-items-center"
    >
      <div
        class="col-auto p-0"
      >
        <button
          class="btn btn-link invisible"
          type="button"
        >
          &lt; Back
        </button>
      </div>
      <div
        class="col-auto"
      >
        <ul>
          <li
            class="invisible"
          />
          <li
            class="invisible"
          />
          <li
            class="invisible"
          />
          <li>
            <span
              class="btn"
              style="cursor: default;"
            >
              1
            </span>
          </li>
          <li>
            <button
              class="btn btn-link"
              type="button"
            >
              2
            </button>
          </li>
          <li>
            <button
              class="btn btn-link"
              type="button"
            >
              3
            </button>
          </li>
          <li>
            <span
              class="btn"
              style="cursor: default;"
            >
              ...
            </span>
          </li>
        </ul>
      </div>
      <div
        class="col-auto p-0"
      >
        <button
          class="btn btn-link"
          type="button"
        >
          More &gt;
        </button>
      </div>
    </div>
  </nav>
</DocumentFragment>
`;

exports[`Pagination renders correctly for whole page range 1`] = `
<DocumentFragment>
  <nav
    aria-label="Search results pages"
    class="custom-pagination container "
  >
    <div
      class="row justify-content-between align-items-center"
    >
      <div
        class="col-auto p-0"
      >
        <button
          class="btn btn-link"
          type="button"
        >
          &lt; Back
        </button>
      </div>
      <div
        class="col-auto"
      >
        <ul>
          <li
            class="invisible"
          />
          <li
            class="invisible"
          />
          <li>
            <button
              class="btn btn-link"
              type="button"
            >
              1
            </button>
          </li>
          <li>
            <span
              class="btn"
              style="cursor: default;"
            >
              2
            </span>
          </li>
          <li>
            <button
              class="btn btn-link"
              type="button"
            >
              3
            </button>
          </li>
          <li>
            <button
              class="btn btn-link"
              type="button"
            >
              4
            </button>
          </li>
          <li
            class="invisible"
          />
        </ul>
      </div>
      <div
        class="col-auto p-0"
      >
        <button
          class="btn btn-link"
          type="button"
        >
          More &gt;
        </button>
      </div>
    </div>
  </nav>
</DocumentFragment>
`;

exports[`Pagination renders correctly in middle of page range 1`] = `
<DocumentFragment>
  <nav
    aria-label="Search results pages"
    class="custom-pagination container "
  >
    <div
      class="row justify-content-between align-items-center"
    >
      <div
        class="col-auto p-0"
      >
        <button
          class="btn btn-link"
          type="button"
        >
          &lt; Back
        </button>
      </div>
      <div
        class="col-auto"
      >
        <ul>
          <li>
            <span
              class="btn"
              style="cursor: default;"
            >
              ...
            </span>
          </li>
          <li>
            <button
              class="btn btn-link"
              type="button"
            >
              3
            </button>
          </li>
          <li>
            <button
              class="btn btn-link"
              type="button"
            >
              4
            </button>
          </li>
          <li>
            <span
              class="btn"
              style="cursor: default;"
            >
              5
            </span>
          </li>
          <li>
            <button
              class="btn btn-link"
              type="button"
            >
              6
            </button>
          </li>
          <li>
            <button
              class="btn btn-link"
              type="button"
            >
              7
            </button>
          </li>
          <li>
            <span
              class="btn"
              style="cursor: default;"
            >
              ...
            </span>
          </li>
        </ul>
      </div>
      <div
        class="col-auto p-0"
      >
        <button
          class="btn btn-link"
          type="button"
        >
          More &gt;
        </button>
      </div>
    </div>
  </nav>
</DocumentFragment>
`;
